src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -19,37 +19,61 @@ public static IEnumerable<TSource> Union<TSource>(this IEnumerable<TSource> firs
         {
             if (first == null) throw Error.ArgumentNull("first");
             if (second == null) throw Error.ArgumentNull("second");
-            return new UnionIterator<TSource>(first, second, comparer);
+            UnionIterator<TSource> union = first as UnionIterator<TSource>;
+            return union != null ? union.Union(second) : new UnionIterator2<TSource>(first, second, comparer);
         }
 
-        private sealed class UnionIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
+        private abstract class UnionIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
         {
-            private readonly IEnumerable<TSource> _first;
-            private readonly IEnumerable<TSource> _second;
-            private readonly IEqualityComparer<TSource> _comparer;
-            private Set<TSource> _set;
+            internal readonly IEqualityComparer<TSource> Comparer;
             private IEnumerator<TSource> _enumerator;
+            private Set<TSource> _set;
 
-            public UnionIterator(IEnumerable<TSource> first, IEnumerable<TSource> second, IEqualityComparer<TSource> comparer)
+            public UnionIterator(IEqualityComparer<TSource> comparer)
             {
-                Debug.Assert(first != null);
-                Debug.Assert(second != null);
-                _first = first;
-                _second = second;
-                _comparer = comparer;
+                Comparer = comparer;
             }
 
-            public override Iterator<TSource> Clone()
+            public override sealed void Dispose()
             {
-                return new UnionIterator<TSource>(_first, _second, _comparer);
+                if (_enumerator != null)
+                {
+                    _enumerator.Dispose();
+                    _enumerator = null;
+                    _set = null;
+                }
+
+                base.Dispose();
             }
 
-            private bool GetNext()
+            internal abstract IEnumerable<TSource> GetSourceEnumerable(int index);
```

Comment:
We should use the same naming as in Concat.  If the preference is for GetSourceEnumerable (I don't have a preference), please change it in Concat as well.